File: pp-exercises/mpi/exercise1/ping-pong/ping-pong.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define MAX_I 20
#define MAX_COUNT (1024 * 1024)

double buf[MAX_COUNT];

int main(int argc, char *argv[])  // ping pong
{
    int size, rank, partner, i, count, r, repetitions;
    double time, bandwidth;
    
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if (size != 2) {
        fprintf(stderr, "comm_size != 2\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
    }

    if (rank == 0) {
        puts("message size    number of         time    bandwidth");
        puts("     [bytes]  repetitions         [us]       [MB/s]");
        puts("---------------------------------------------------");
    }
                     
    partner = 1 - rank;
    count = 1;
    repetitions = MAX_COUNT;

    for (i = 0; i <= MAX_I; i++) {
        time = MPI_Wtime();
        for (r = 0; r < repetitions; r++) {
            if (rank == 0) {
                MPI_Send(buf, count, MPI_DOUBLE, partner, 0, MPI_COMM_WORLD);
                MPI_Recv(buf, count, MPI_DOUBLE, partner, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            } else {
                MPI_Recv(buf, count, MPI_DOUBLE, partner, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
                MPI_Send(buf, count, MPI_DOUBLE, partner, 0, MPI_COMM_WORLD);
            }
        }
        time = MPI_Wtime() - time;
        time = time / (2.0 * repetitions);
        time = 1e6 * time;  // micro-seconds
        bandwidth = count * sizeof(double) / time;  // MByte/s

        if (rank == 0)
            printf("%12zd %12d %12.2f %12.2f\n", count * sizeof(double), repetitions, time, bandwidth);
        count = count * 2;
        repetitions = repetitions / 2;
    }

    MPI_Finalize();
    return 0;
}  

File: pp-exercises/mpi/exercise1/ping-pong/ping-pong.f90
program main  ! ping pong

  implicit none
  include 'mpif.h'
  integer :: size, rank, ierr, partner, i, count, r, repetitions
  integer, parameter :: max_i = 20
  integer, parameter :: max_count = 2**max_i
  real(8) :: buf(max_count), time, bandwidth

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  if (size /= 2) then
     write(0,*) "comm_size != 2"
     call mpi_abort(MPI_COMM_WORLD, 1, ierr)
  endif

  if (rank == 0) then
     write(6,"(a)") "message size    number of         time    bandwidth"
     write(6,"(a)") "     [bytes]  repetitions         [us]       [MB/s]"
     write(6,"(a)") "---------------------------------------------------"
  endif
  
  partner = 1 - rank
  count = 1
  repetitions = max_count

  do i = 0, max_i
     time = mpi_wtime()
     do r = 1, repetitions
        if (rank == 0) then
           call mpi_send(buf, count, MPI_REAL8, partner, 0, MPI_COMM_WORLD, ierr)
           call mpi_recv(buf, count, MPI_REAL8, partner, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
        else
           call mpi_recv(buf, count, MPI_REAL8, partner, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
           call mpi_send(buf, count, MPI_REAL8, partner, 0, MPI_COMM_WORLD, ierr)
        endif
     enddo
     time = mpi_wtime() - time
     time = time / (2.0 * repetitions)
     time = 1e6 * time  ! micro-seconds
     bandwidth = count * 8 / time  ! MByte/s
     
     if (rank == 0) write(6,"(i12,x,i12,2(x,f12.2))") count * 8, repetitions, time, bandwidth
     count = count * 2
     repetitions = repetitions / 2
  enddo

  call mpi_finalize(ierr)
end
